src/components/parser.cpp

```diff
@@ -33,15 +33,29 @@ parser::parser(signal_emitter& emitter) : m_sig(emitter) {}
  */
 void parser::parse(const bar_settings& bar, string data) {
   while (!data.empty()) {
-    size_t pos{string::npos};
-
-    if (data.compare(0, 2, "%{") == 0 && (pos = data.find('}')) != string::npos) {
-      codeblock(data.substr(2, pos - 2), bar);
-      data.erase(0, pos + 1);
-    } else if ((pos = data.find("%{")) != string::npos) {
-      data.erase(0, text(data.substr(0, pos)));
+    size_t beg_pos{string::npos};
```

Comment:
Why was it necessary to rewrite the parser? This feature should really only affect the string produced by the builder.

Also, we use `%` to escape `%{` blocks because that's the way lemonbar does it. This would also break existing configs relying on this escaping behavior.

Reply:
Hum, maybe I missed something but the actual parser does not escape when there is a '%'.
I used '\' because I could not differentiate if the following string was tag escaping or just a percentage and a tag: '97%%{O100}'. But maybe I can double each percentage when a label is created.

Reply:
> the actual parser does not escape when there is a '%'.

Oh, that seems like an oversight, it really should. I was kind of dreading this moment. It is entirely possible that the whole escaping business inside polybar breaks in various places if someone is taking a closer look. I was taking a look at the parser a few months back but couldn't make heads or tails of what polybar does with escaping and I just wrote down all my questions and put them away for later. Seems my question "How is escaping of formatting tags handled?" was answered: "Not at all".

I'm not really sure if it is even possible to escape to properly detect `97%%{O100}` without completely parsing all user supplied raw formatting tags. If you come up with a solution, let me know. Maybe wait a bit with implementing it so we can discuss it first.

Reply:
I think that's indeed impossible. As we discussed on the IRC, maybe the best is just breaking the backward compatibility and using the comportment of Lemonbar.

I will rewrite the parser to take in account that '%' is a special character. And as you said, this PR should not be merged before the 4.0 of polybar.

However, maybe we can use the solution proposed in #1065 (with random IDs) as a temporary solution while waiting the 4.0.

Reply:
I think we have to use those X tags. But maybe we can build those on top of https://github.com/jaagr/polybar/pull/1596 which introduces the `%{P` tag for polybar control commands, so we could use `%{PX`

Reply:
Looking at this again, and man this is a difficult problem.

I think the most elegant way to solve this (and also the most time-consuming) is to stop using lemonbar formatting tags internally (at least in string form). So instead of passing a large formatting string from the controller to the parser and renderer, we pass an array of formatting elements. This array is basically the formatting string represented using C++ data structures, so each element is either just plain text or describes a formatting tag.

Formatting tags are parsed when text is read from labels, formats, ramps, tokens, etc. From then on out we never deal with formatting tags as pure strings anymore.

I think that can solve this issue without breaking too many configs. The configs that will still break are the ones that literally write something like `97%%{O100}` in one of their labels or as part of the script output. I think this is acceptable even without a major release.